cmd/hydrate/src/lib.rs

```diff
@@ -0,0 +1,217 @@
+    unsupported!(reset_and_halt(_dur: std::time::Duration));
+    unsupported!(wait_for_halt(_dur: std::time::Duration));
+
+    fn info(&self) -> (String, Option<String>) {
+        ("DryCore".to_owned(), None)
+    }
+    fn is_archive(&self) -> bool {
+        false
+    }
+    fn is_dump(&self) -> bool {
+        true // I guess?
+    }
+    fn is_net(&self) -> bool {
+        false
+    }
+    fn vid_pid(&self) -> Option<(u16, u16)> {
+        None
+    }
+
+    fn read_8(&mut self, addr: u32, data: &mut [u8]) -> Result<()> {
+        if self.flash.read(addr, data).is_some() {
+            return Ok(());
+        }
+
+        let Some((base, mem)) = self.mem.range(0..=addr).next_back() else {
+            bail!("addr {addr:#08x} is below memory range");
+        };
+        let offset = (addr - base) as usize;
+        let end = offset + data.len();
+        if end > mem.len() {
+            bail!("region is not large enough; {end:#x} > {:#x}", mem.len());
+        }
+        data.copy_from_slice(&mem[offset..end]);
+        Ok(())
+    }
+}
+
+fn run(context: &mut ExecutionContext) -> Result<()> {
+    let Subcommand::Other(subargs) = context.cli.cmd.as_ref().unwrap();
+    let subargs = Args::try_parse_from(subargs)?;
+    let f = std::fs::File::open(&subargs.file)?;
+    let mut z = zip::ZipArchive::new(f)?;
+
+    let mut s = String::new();
+    z.by_name("meta.json")?.read_to_string(&mut s)?;
+    #[derive(serde::Deserialize)]
+    struct Meta {
+        version: u64,
+    }
+    let meta: Meta = serde_json::from_str(&s)?;
+    if meta.version != 1 {
+        bail!(
+            "invalid version in `meta.json`: expected 1, got {}",
+            meta.version
+        );
+    }
+
+    let mut s = String::new();
+    z.by_name("TASK_INDEX")?.read_to_string(&mut s)?;
+    let task_id: u16 = s.trim().parse()?;
+
+    let mut s = String::new();
+    z.by_name("TIMESTAMP")?.read_to_string(&mut s)?;
+    let timestamp: u64 = s.trim().parse()?;
+
+    let mut bord = String::new();
+    z.by_name("BORD")?.read_to_string(&mut bord)?;
+    bord.pop(); // remove newline
+
+    let mut gitc = String::new();
+    z.by_name("GITC")?.read_to_string(&mut gitc)?;
+    gitc.pop(); // remove newline
+
+    let vers = match z.by_name("VERS") {
+        Ok(mut v) => {
+            let mut s = String::new();
+            v.read_to_string(&mut s)?;
+            s.pop();
+            Some(s)
+        }
+        Err(zip::result::ZipError::FileNotFound) => None,
+        Err(e) => return Err(e.into()),
+    };
+
+    let mut s = String::new();
+    z.by_name("ARCHIVE_ID")?.read_to_string(&mut s)?;
+    let mut archive_id = [0u8; 8];
+    for (i, b) in archive_id.iter_mut().enumerate() {
+        *b = u8::from_str_radix(&s[i * 2..][..2], 16)?;
+    }
+
+    let mut mem = BTreeMap::new();
+    let mem_files = z
+        .file_names()
+        .filter(|f| f.starts_with("0x"))
+        .map(|s| s.to_owned())
+        .collect::<Vec<_>>();
```

Comment:
~~nit, not a big deal: we should be able to drop the `collect` here, and instead just `for` loop over the iterator directly, right?~~

EDIT: oh, I bet this is necessary because the iterator borrows the `ZipArchive`, but calling `ZipArchive::by_name` inside the loop mutably borrows it, so we can't do that. Maybe worth a comment here?